sw/host/opentitanlib/src/transport/cw310/usb.rs

```diff
@@ -0,0 +1,546 @@
+            &[sdi, Backend::CONFIG_PIN_SPI1_SDI],
+        )?;
+        self.usb.send_ctrl(
+            Backend::CMD_FPGAIO_UTIL,
+            Backend::REQ_IO_CONFIG,
+            &[sck, Backend::CONFIG_PIN_SPI1_SCK],
+        )?;
+        self.usb.send_ctrl(
+            Backend::CMD_FPGAIO_UTIL,
+            Backend::REQ_IO_CONFIG,
+            &[cs, Backend::CONFIG_PIN_SPI1_CS],
+        )?;
+        Ok(())
+    }
+
+    /// Enable the spi interface on the SAM3U chip.
+    pub fn spi1_enable(&self, enable: bool) -> Result<()> {
+        self.usb.send_ctrl(
+            Backend::CMD_FPGASPI1_XFER,
+            if enable {
+                Backend::REQ_ENABLE_SPI
+            } else {
+                Backend::REQ_DISABLE_SPI
+            },
+            &[],
+        )?;
+        Ok(())
+    }
+
+    /// Set the value of the SPI chip-select pin.
+    pub fn spi1_set_cs_pin(&self, status: bool) -> Result<()> {
+        self.usb.send_ctrl(
+            Backend::CMD_FPGASPI1_XFER,
+            if status {
+                Backend::REQ_CS_HIGH
+            } else {
+                Backend::REQ_CS_LOW
+            },
+            &[],
+        )?;
+        Ok(())
+    }
+
+    /// Perform an up to 64-byte transfer on the SPI interface.
+    /// This is a low-level function and does not affect the CS pin.
+    pub fn spi1_tx_rx(&self, txdata: &[u8], rxdata: &mut [u8]) -> Result<()> {
+        ensure!(
+            txdata.len() <= 64,
+            SpiError::InvalidDataLength(txdata.len())
+        );
+        ensure!(
+            rxdata.len() <= 64,
+            SpiError::InvalidDataLength(rxdata.len())
+        );
+        ensure!(
+            txdata.len() == rxdata.len(),
+            SpiError::MismatchedDataLength(txdata.len(), rxdata.len())
+        );
+        self.usb
+            .send_ctrl(Backend::CMD_FPGASPI1_XFER, Backend::REQ_SEND_DATA, txdata)?;
+        self.usb.read_ctrl(Backend::CMD_FPGASPI1_XFER, 0, rxdata)?;
+        Ok(())
+    }
+
+    /// Read a `buffer` slice from the SPI interface.
+    /// This is a low-level function and does not affect the CS pin.
+    pub fn spi1_read(&self, buffer: &mut [u8]) -> Result<()> {
+        let wbuf = [0u8; 64];
+        for chunk in buffer.chunks_mut(64) {
+            self.spi1_tx_rx(&wbuf[..chunk.len()], chunk)?;
+        }
+        Ok(())
+    }
+
+    /// Write a `buffer` slice to the SPI interface.
+    /// This is a low-level function and does not affect the CS pin.
+    pub fn spi1_write(&self, buffer: &[u8]) -> Result<()> {
+        let mut rbuf = [0u8; 64];
+        for chunk in buffer.chunks(64) {
+            self.spi1_tx_rx(chunk, &mut rbuf[..chunk.len()])?;
+        }
+        Ok(())
+    }
+
+    /// Perform a write & read transaction to the SPI interface.
+    /// This is a low-level function and does not affect the CS pin.
+    pub fn spi1_both(&self, txbuf: &[u8], rxbuf: &mut [u8]) -> Result<()> {
+        for (wchunk, rchunk) in txbuf.chunks(64).zip(rxbuf.chunks_mut(64)) {
```

Comment:
Note that if txbuf and rxbuf are not within each others' size, it will exit early; zip short-circuits upon either of the iterators returning None. Is this intentional?

Reply:
I added an explicit check in this function.  I think it would have been caught by the `spi1_tx_rx` function, but it's probably better to error out before even starting the SPI transaction.